src/connectors/pagination/connectPagination.ts

```diff
@@ -0,0 +1,191 @@
+ */
+const connectPagination: PaginationConnector = (renderFn, unmountFn = noop) => {
+  checkRendering(renderFn, withUsage());
+
+  return widgetParams => {
+    const { totalPages, padding = 3 } =
+      widgetParams || ({} as PaginationConnectorParams);
+
+    const pager = new Paginator({
+      currentPage: 0,
+      total: 0,
+      padding,
+    });
+
+    type ConnectorState = {
+      refine?: (page: number) => void;
+      createURL?: (state: SearchParameters) => (page: number) => string;
+    };
+
+    const connectorState: ConnectorState = {};
+
+    const getMaxPage = ({ nbPages }) =>
+      totalPages !== undefined ? Math.min(totalPages, nbPages) : nbPages;
+
+    return {
+      $$type: 'ais.pagination',
+
+      init({ helper, createURL, instantSearchInstance }) {
+        connectorState.refine = page => {
+          helper.setPage(page);
+          helper.search();
+        };
+
+        connectorState.createURL = state => page =>
+          createURL(state.setPage(page));
+
+        renderFn(
+          {
+            createURL: connectorState.createURL!(helper.state),
+            currentRefinement: helper.state.page || 0,
+            nbHits: 0,
+            nbPages: 0,
+            pages: [],
+            isFirstPage: true,
+            isLastPage: true,
+            refine: connectorState.refine!,
+            widgetParams,
+            instantSearchInstance,
+          },
+          true
+        );
+      },
+
+      render({ results, state, instantSearchInstance }) {
+        const page = state.page || 0;
+        const nbPages = getMaxPage(results);
+        pager.currentPage = page;
+        pager.total = nbPages;
+
+        renderFn(
+          {
+            createURL: connectorState.createURL!(state),
+            currentRefinement: page,
+            refine: connectorState.refine!,
+            nbHits: results.nbHits,
+            nbPages,
+            pages: pager.pages(),
+            isFirstPage: pager.isFirstPage(),
+            isLastPage: pager.isLastPage(),
+            widgetParams,
+            instantSearchInstance,
+          },
+          false
+        );
+      },
+
+      dispose({ state }) {
+        unmountFn();
+
+        return state.setQueryParameter('page', undefined);
+      },
+
+      getWidgetState(uiState, { searchParameters }) {
+        const page = searchParameters.page || 0;
+
+        if (!page) {
+          return uiState;
+        }
+
+        return {
+          ...uiState,
+          page: page + 1,
+        };
+      },
+
+      getWidgetSearchParameters(searchParameters, { uiState }) {
+        const page = uiState.page ? uiState.page - 1 : 0;
+
+        return searchParameters.setQueryParameter('page', page);
+      },
+
+      // for tests, remove when refactored
+      refine: page => connectorState.refine!(page),
+      createURL: helper => connectorState.createURL!(helper),
+      getMaxPage,
```

Comment:
I would have prefered we include the refactoring of the tests before merging, because I sense it's the kind of thing we end up never doing.
Is there a reason (other than PR bloat) not to include them in this diff?

Reply:
PR bloat and I wasn’t sure what the alternative way of testing this could be.

Reply:
I agree with @tkrugg.

There's no need to test `widget.refine()` at the widget level. We should rather mount the widget and interact with it to trigger the expected behaviors.